File: HLS/perf_opt1/peakPicker.m
% Implementation of a hardware-friendly peak detection algorithm
% Computing CompareOut = current_window - MidSample
% Using all(CompareOut <= 0) to check if the middle sample is a local maximum
% Simplified conditional logic
% Modified by Jie Lei, University of Technology Sydney, 2025

function [locations]= peakPicker(xcorr,threshold,window_length)
% Copyright 2021-2023 The MathWorks, Inc.

    locations=[];
    middle_location=floor(window_length/2);
    
    for index=1:length(xcorr)-window_length+1
        current_window=xcorr(index:index+window_length-1,:); % sliding window
        candidate_location=index+middle_location;
        % Hardware friendly implementation of peak finder
        MidSample = current_window(middle_location+1,:);
        CompareOut = current_window - MidSample; % this is a vector
        % if all values in the result are negative and the middle sample is
        % greater than a threshold, it is a local max
        if all(CompareOut <= 0) && (MidSample > threshold(candidate_location))
            locations = [locations candidate_location]; %#ok
        end
    end
end

File: HLS/perf_opt1/peakPicker.cpp
#include "peakPicker.hpp"

void peakPicker(
    hls::stream<DataType>& xcorrStream,
    hls::stream<DataType>& thresholdStream,
    hls::stream<LocationType>& locationsStream,
    uint16_t signalLength,
    uint16_t windowLength,
    uint16_t& numPeaks
) {
    #pragma HLS INTERFACE mode=axis port=xcorrStream
    #pragma HLS INTERFACE mode=axis port=thresholdStream
    #pragma HLS INTERFACE mode=axis port=locationsStream
    #pragma HLS INTERFACE mode=s_axilite port=signalLength bundle=control
    #pragma HLS INTERFACE mode=s_axilite port=windowLength bundle=control
    #pragma HLS INTERFACE mode=s_axilite port=numPeaks bundle=control
    #pragma HLS INTERFACE mode=s_axilite port=return bundle=control

    // Buffer for input data
    DataType xcorr[MAX_SIGNAL_LENGTH];
    #pragma HLS RESOURCE variable=xcorr core=RAM_2P_BRAM
    
    // Buffer for threshold data
    DataType threshold[MAX_SIGNAL_LENGTH];
    #pragma HLS RESOURCE variable=threshold core=RAM_2P_BRAM
    
    // Read input data from streams
    InputRead: for (uint16_t i = 0; i < signalLength; i++) {
        #pragma HLS PIPELINE II=1
        xcorr[i] = xcorrStream.read();
        threshold[i] = thresholdStream.read();
    }
    
    // Calculate middle location of the window
    uint16_t middleLocation = windowLength / 2;
    
    // Reset peak counter
    numPeaks = 0;
    
    // Process the signal to find peaks
    ProcessSignal: for (uint16_t index = 0; index < signalLength - windowLength + 1; index++) {
        #pragma HLS PIPELINE II=1
        
        // Get the candidate location (middle of current window)
        uint16_t candidateLocation = index + middleLocation;
        
        // Get the middle sample value
        DataType midSample = xcorr[candidateLocation];
        
        // Check if the middle sample is a local maximum
        bool isLocalMax = true;
        
        WindowCheck: for (uint16_t w = 0; w < windowLength; w++) {
            #pragma HLS UNROLL
            if (w != middleLocation) {
                DataType windowSample = xcorr[index + w];
                if (windowSample > midSample) {
                    isLocalMax = false;
                }
            }
        }
        
        // Check if the middle sample exceeds the threshold
        bool exceedsThreshold = (midSample > threshold[candidateLocation]);
        
        // If both conditions are met, we have a peak
        if (isLocalMax && exceedsThreshold && numPeaks < MAX_PEAKS) {
            locationsStream.write(candidateLocation+1); // +1 for 1-based index
            numPeaks++;
        }
    }
}
